fragment.go

```diff
@@ -218,24 +219,29 @@ func (f *fragment) openStorage() error {
 		if err != nil {
 			return errors.Wrap(err, "statting file after")
 		}
-	}
-
-	// Mmap the underlying file so it can be zero copied.
-	storageData, err := syscall.Mmap(int(f.file.Fd()), 0, int(fi.Size()), syscall.PROT_READ, syscall.MAP_SHARED)
-	if err != nil {
-		return fmt.Errorf("mmap: %s", err)
-	}
-	f.storageData = storageData
+	} else {
+		// Mmap the underlying file so it can be zero copied.
+		data, err := syswrap.Mmap(int(f.file.Fd()), 0, int(fi.Size()), syscall.PROT_READ, syscall.MAP_SHARED)
+		if err == syswrap.ErrMaxMapCountReached {
+			f.Logger.Debugf("maximum number of maps reached, reading file instead")
+			data, err = ioutil.ReadAll(file)
```

Comment:
This doesn't seem insane, but I wonder whether we should, in this case, unmap all the containers so we can drop the data object, since that's likely a large allocation.

Reply:
I think this would not be trivial at the moment—since unmapping would mean copying everything. Would be nice to have some kind of streaming variant of unmarshalbinary that creates unmapped containers, but I'm not sure it would ever be as performant as doing pointer magic into a big byte slice.

My future goals for this would be to have the ability to only read the header, and then fetch rows from storage on-demand and cache them.

Or something like that... more thinking needed.